### PR TITLE
Che multi user deploy to openshift

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?

This PR proposes some changes required to enable the `che-server-multiuser` docker image to run on Minishift, as well as the required Postgres database server and optionally the customized Keycloak server. 

### How to test multi-user CHE on Minishift without building anything:

You can test it through already build docker images with the following steps:

```
# You can set  CHE_IMAGE_REPO and CHE_IMAGE_TAG to your locally-built Che multi-server docker image.
export CHE_IMAGE_REPO="<YourDockerAccount>/<YourImageName>"
export CHE_IMAGE_TAG="<YourImageTag>"

export CHE_MULTI_USER=true

docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v $(pwd)/multiuser-conf:/data -e IMAGE_INIT=dfestal/che-init:nightly -e CHE_MULTIUSER=true eclipse/che-cli:nightly config

./multiuser-conf/instance/config/openshift/scripts/deploy_che.sh
``` 

### To build everything locally and start multi-user Che on Minishift:

- Rebuild upstream che from the `davidfestal:che-multi-user-deploy-to-openshift`
- Check that Minishift is running
- Connect the current terminal to the Minishift docker daemon
- Tag your local `che-init` image with you docker.io account and push it:
```
docker tag eclipse/che-init:nightly docker.io/<YourDockerAccount>/che-init:nightly
docker push docker.io/<YourDockerAccount>/che-init:nightly
```
- Rebuild `init`, `cli` and `che` images (in the given sequence). To do that, proceed to folder `dockerfiles/<image_name>` and run `build.sh`
- Change directory to `dockerfiles/init/modules/openshift/files/scripts`
- Set the following environment variables:
```
  export CHE_MULTI_USER=true
  export CHE_DEDICATED_KEYCLOAK=true
  export IMAGE_INIT=<YourDockerAccount>/che-init:nightly
```
- run the following commands:
```
./deploy_che.sh
```

### CODE REVIEW

**Comment by benoitf** [comment]
ci-build

File: dockerfiles/cli/scripts/openshift/deploy_che.sh
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/deploy_che.sh]: i'm not sure we need to add this check as part of the switch on the flavor, it can be before the switch

File: dockerfiles/cli/scripts/openshift/deploy_che.sh
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/deploy_che.sh]: on a global remark, this script should not be in cli/scripts but probably in base/scripts and fit the existing model on how to define commands

File: dockerfiles/cli/scripts/openshift/deploy_che.sh
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/deploy_che.sh]: using set +e / set -e looks like a hack
why not using the `||` fallback ?

File: dockerfiles/cli/scripts/openshift/deploy_che.sh
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/deploy_che.sh]: we should be always in set -e mode

File: dockerfiles/cli/scripts/openshift/deploy_che.sh
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/deploy_che.sh]: could we use a function taking a name and value ?
looks like it's hard to maintain.

File: dockerfiles/cli/scripts/openshift/multi-user/che-init-image-stream.yaml
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/multi-user/che-init-image-stream.yaml]: all .yaml files need to be part of che-init image.

File: dockerfiles/cli/scripts/openshift/multi-user/che-init-image-stream.yaml
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/multi-user/che-init-image-stream.yaml]: there is a missing newline on this file and many others

File: dockerfiles/cli/scripts/openshift/multi-user/che-init-image-stream.yaml
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/multi-user/che-init-image-stream.yaml]: missing copyright headers

File: dockerfiles/cli/scripts/openshift/multi-user/deployPostgresAndKeycloak.sh
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/multi-user/deployPostgresAndKeycloak.sh]: the name of the file is not following the scripts used by the cli at all

File: dockerfiles/cli/scripts/openshift/multi-user/deployPostgresAndKeycloak.sh
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/multi-user/deployPostgresAndKeycloak.sh]: IMHO we shouldn't have any hardcoded value of image in a script

File: dockerfiles/cli/scripts/openshift/multi-user/deployPostgresOnly.sh
Comment (benoitf) [inline dockerfiles/cli/scripts/openshift/multi-user/deployPostgresOnly.sh]: we're missing some utilities function like for creating something with oc create

**Comment by benoitf** [comment]
Could you use https://www.shellcheck.net/ to verify all your shell scripts ?

**Review by benoitf — DISMISSED** [review]

File: dockerfiles/cli/scripts/openshift/deploy_che.sh
Comment (davidfestal) [reply dockerfiles/cli/scripts/openshift/deploy_che.sh]: I moved it, as well as the other scripts and openshift descriptors to `che-init`

File: dockerfiles/init/modules/compose/templates/docker-compose.yml.erb
Comment (benoitf) [inline dockerfiles/init/modules/compose/templates/docker-compose.yml.erb]: is that timeout should be configurable by end users with ENV value ?

Comment (davidfestal) [reply dockerfiles/init/modules/compose/templates/docker-compose.yml.erb]: That might be yes, but maybe in a next PR ?

File: dockerfiles/cli/images.template
Comment (davidfestal) [reply dockerfiles/cli/images.template]: Sure. In fact this correspond to the following PR https://github.com/jboss-dockerfiles/keycloak/pull/79. I'm currently waiting for an answer about whether / when this should be merged and made available as `jboss/keycloak-postgres-openshift:3.2.1.Final` in in DockerHub.

Maybe we could still merge this PR to the `che-multiuser` branch, and later ensure we change this to the final image name before merging the `che-multiuser` branch to master ?

File: dockerfiles/init/modules/compose/templates/docker-compose.yml.erb
Comment (riuvshin) [inline dockerfiles/init/modules/compose/templates/docker-compose.yml.erb:116]: why pg_isready does not work here?

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/che-init-image-stream.yaml
Comment (riuvshin) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/che-init-image-stream.yaml]: new line

File: dockerfiles/init/modules/postgres/templates/init-che-user.sh.erb
Comment (riuvshin) [inline dockerfiles/init/modules/postgres/templates/init-che-user.sh.erb]: new line

**Comment by davidfestal** [comment]
> Could you use https://www.shellcheck.net/ to verify all your shell scripts ?

@benoitf I've done it and made the requested  changes (apart one which wasn't related to my PR)

File: dockerfiles/init/modules/postgres/templates/init-che-user.sh.erb
Comment (davidfestal) [reply dockerfiles/init/modules/postgres/templates/init-che-user.sh.erb]: fixed

File: dockerfiles/init/modules/compose/templates/docker-compose.yml.erb
Comment (davidfestal) [reply dockerfiles/init/modules/compose/templates/docker-compose.yml.erb:116]: This PR now uses the official, CentOS-based, Postgres docker image of the Openshift Online catalog (https://console.starter-us-east-2.openshift.com/console/project/dfestal/create/fromtemplate?template=postgresql-persistent&namespace=openshift), whose sources are here: https://github.com/sclorg/postgresql-container/tree/master/9.5.
This image is compatible with *bothù simple docker and Openshift deployments, which is certainly a good thing for maintenance.

Since it's based on CentOS and uses SCL, `pg_isready` is not directly available for the health check. Additionaly, I thought it was a good thing to use the same health-check as the one officially used for this image on Openshift Online, as defined here: https://github.com/openshift/library/blob/master/official/postgresql/templates/postgresql-persistent.json#L159

**Comment by davidfestal** [comment]
@benoitf is it OK for you to approve this PR now, to have it merge to the `che-multiuser` branch only, or is there any other blocker issue ?

**Comment by davidfestal** [comment]
@riuvshin is it OK for you ?

**Comment by davidfestal** [comment]
@skabashnyuk @mshaposhnik do you think you could review this PR ? It would be quite helpful for the remaining work on multi-user subject, to have it merged into the main `che-multiuser` branch.
That would allow you to test your ongoing work very easily on Minishift.

File: dockerfiles/cli/images.template
Comment (skabashnyuk) [reply dockerfiles/cli/images.template]: @riuvshin can you suggest what is the better way for you?

**Review by skabashnyuk — COMMENTED** [review]
Up to @riuvshin . It's more his part PR.
+0

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/deployPostgresAndKeycloak.sh
Comment (benoitf) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/deployPostgresAndKeycloak.sh]: could we avoid to use Uppercase in the name of the scripts ?
we have only `lower hyphen` or `lower_underscore`  sh files

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/deployPostgresOnly.sh
Comment (benoitf) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/deployPostgresOnly.sh]: do we need to hardcode the default name of the image there ? it may lead to unexpected errors

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/deployPostgresOnly.sh
Comment (benoitf) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/deployPostgresOnly.sh]: do we need to hardcode the default name of the image there ? it may lead to unexpected errors

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/keycloak-data-claim.yaml
Comment (benoitf) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/keycloak-data-claim.yaml]: missing newline

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/service.yaml
Comment (benoitf) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/service.yaml]: missing newline

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/postgres/deployment-config.yaml
Comment (benoitf) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/postgres/deployment-config.yaml]: extra newline ? :-)

File: dockerfiles/init/s2i/keycloak/assemble
Comment (benoitf) [inline dockerfiles/init/s2i/keycloak/assemble:1]: missing headers

File: dockerfiles/init/s2i/keycloak/run
Comment (benoitf) [inline dockerfiles/init/s2i/keycloak/run]: missing copyrights

File: dockerfiles/init/s2i/postgres/assemble
Comment (benoitf) [inline dockerfiles/init/s2i/postgres/assemble:1]: missing copyrights

File: dockerfiles/init/s2i/postgres/run
Comment (benoitf) [inline dockerfiles/init/s2i/postgres/run]: missing copyrights

File: dockerfiles/cli/scripts/openshift/deploy_che.sh
Comment (l0rd) [reply dockerfiles/cli/scripts/openshift/deploy_che.sh]: @benoitf @davidfestal I would prefer if we you keep this script where it is. This PR is not about fixing the location of `deploy_che.sh`. Fixing that will be handled in another PR on master branch. Moving `deploy_che.sh` to another folder in this PR will make merging `multi-user` branch to `master` more complicated.

File: dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh
Comment (l0rd) [inline dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh]: Use ${CHE_MULTI_USER} instead of $CHE_MULTI_USER to be more consistent with coding style in the rest of this script.

File: dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh
Comment (l0rd) [inline dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh]: Why are you changing that?

File: dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh
Comment (l0rd) [inline dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh]: Please be more specific so that someone that is not @davidfestal can understand :-). Something like (not sure I've understand what need to be done but...):
```bash
# TODO move this env variable as a config map in the deployment config as soon as che-multiuser branch is merged to master
```

File: dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh
Comment (l0rd) [reply dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh]: And `then` keyword should be on the same line of the test for the same reason.

Comment (l0rd) [reply dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh]: The same applies to other if conditions you have added in the script

File: dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh
Comment (l0rd) [inline dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh]: Please prefix this with `[CHE] **ERROR**: ` to be consistent with the rest of the of the script and make it clear that there is an error.

Comment (l0rd) [reply dockerfiles/init/modules/openshift/files/scripts/deploy_che.sh]: The same applies to other `echo` you have added.

File: dockerfiles/cli/scripts/openshift/deploy_che.sh
Comment (davidfestal) [reply dockerfiles/cli/scripts/openshift/deploy_che.sh]: The PR to move the `deploy_che.sh` script is under review on master  (#6390). As soon as it is merged, and `master` is merged into the `che-multiuser` branch, I'll rebase this PR on `che-multiuser` and things will be clean.

Is it OK for you @l0rd ?

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/keycloak-data-claim.yaml
Comment (davidfestal) [reply dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/keycloak-data-claim.yaml]: Fixed afaik

File: dockerfiles/init/s2i/keycloak/assemble
Comment (davidfestal) [reply dockerfiles/init/s2i/keycloak/assemble:1]: fixed

File: dockerfiles/init/s2i/postgres/assemble
Comment (davidfestal) [reply dockerfiles/init/s2i/postgres/assemble:1]: fixed

File: dockerfiles/cli/images.template
Comment (benoitf) [reply dockerfiles/cli/images.template]: @riuvshin  ?

Comment (riuvshin) [reply dockerfiles/cli/images.template]: +1 that should be changed to official image before merge

File: dockerfiles/cli/version/latest/images
Comment (riuvshin) [inline dockerfiles/cli/version/latest/images]: same, please use official image

File: dockerfiles/init/modules/compose/templates/docker-compose.yml.erb
Comment (riuvshin) [inline dockerfiles/init/modules/compose/templates/docker-compose.yml.erb:148]: does this change a way how realms are added? because AFAIK keycloak use /export folder to export configs which work only once. if we mount our files to `/opt/jboss/keycloak/realms/` does this will change that behavior? I mean each further change will also applied not as it was with `/export`

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/deploy_postgres_and_keycloak.sh
Comment (riuvshin) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/deploy_postgres_and_keycloak.sh]: please use official image, btw do we need that here duplicated?

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/deployment-config.yaml
Comment (riuvshin) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/deployment-config.yaml]: maybe we need somehow make this configurable? as I understand this is hardcoded and what should we do in case of change these values? (if don't need change it at all then it;s ok)

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/postgres/build-config.yaml
Comment (riuvshin) [inline dockerfiles/init/modules/openshift/files/scripts/multi-user/postgres/build-config.yaml:40]: is it OK to use always `latest` tag here? I have no idea how these yamls will be used but I think that could be a problem to use always latest.

File: dockerfiles/init/modules/postgres/manifests/init.pp
Comment (riuvshin) [inline dockerfiles/init/modules/postgres/manifests/init.pp:4]: 26 is pg user id ?

Comment (davidfestal) [reply dockerfiles/init/modules/postgres/manifests/init.pp:4]: yes

File: dockerfiles/init/modules/keycloak/templates/che-realm.json.erb
Comment (skabashnyuk) [inline dockerfiles/init/modules/keycloak/templates/che-realm.json.erb]: Do we have one client here or two? ```che-public```  or ```che``` and ```che-public```?

File: dockerfiles/init/modules/compose/templates/docker-compose.yml.erb
Comment (davidfestal) [reply dockerfiles/init/modules/compose/templates/docker-compose.yml.erb:148]: No, it doesn't since overriding previous realms is driven by the following parameter:

    -Dkeycloak.migration.strategy=IGNORE_EXISTING

The first time I deploy the Keycloak server (with an empty persistent volume), I see these messages in the start logs:
```
11:15:19,806 INFO  [org.keycloak.exportimport.dir.DirImportProvider] (ServerService Thread Pool -- 50) Importing from directory /opt/jboss/keycloak/realms
--
  | 11:15:19,819 INFO  [org.keycloak.services] (ServerService Thread Pool -- 50) KC-SERVICES0030: Full model import requested. Strategy: IGNORE_EXISTING
  | 11:15:23,036 INFO  [org.keycloak.exportimport.util.ImportUtils] (ServerService Thread Pool -- 50) Realm 'master' imported
  | 11:15:23,910 INFO  [org.keycloak.exportimport.dir.DirImportProvider] (ServerService Thread Pool -- 50) Imported users from /opt/jboss/keycloak/realms/master-users-0.json
  | 11:15:25,807 INFO  [org.keycloak.exportimport.util.ImportUtils] (ServerService Thread Pool -- 50) Realm 'che' imported
  | 11:15:26,070 INFO  [org.keycloak.exportimport.dir.DirImportProvider] (ServerService Thread Pool -- 50) Imported users from /opt/jboss/keycloak/realms/che-users-0.json
  | 11:15:26,146 INFO  [org.keycloak.services] (ServerService Thread Pool -- 50) KC-SERVICES0032: Import finished successfully
```

But if I restart the keycloak server (with its data already available in the persistent volume), I see the following message in the logs: 
```
11:36:15,183 INFO  [org.keycloak.exportimport.dir.DirImportProvider] (ServerService Thread Pool -- 60) Importing from directory /opt/jboss/keycloak/realms
--
  | 11:36:16,264 INFO  [org.keycloak.services] (ServerService Thread Pool -- 60) KC-SERVICES0030: Full model import requested. Strategy: IGNORE_EXISTING
  | 11:36:16,744 INFO  [org.keycloak.exportimport.util.ImportUtils] (ServerService Thread Pool -- 60) Realm 'master' already exists. Import skipped
  | 11:36:16,950 INFO  [org.keycloak.exportimport.util.ImportUtils] (ServerService Thread Pool -- 60) Realm 'che' already exists. Import skipped
  | 11:36:16,998 INFO  [org.keycloak.services] (ServerService Thread Pool -- 60) KC-SERVICES0032: Import finished successfully
```

And I can confirm that the settings done in the Keycloak adm console are kept after restart.

File: dockerfiles/init/modules/keycloak/templates/che-realm.json.erb
Comment (davidfestal) [reply dockerfiles/init/modules/keycloak/templates/che-realm.json.erb]: I haven't rebased on your last commits where you removed one of the clients.
I'll do it just after and one the clients will probably be removed

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/postgres/build-config.yaml
Comment (davidfestal) [reply dockerfiles/init/modules/openshift/files/scripts/multi-user/postgres/build-config.yaml:40]: yes, because the `che-init:latest` image stream tag is itself generated from the configured  docker `che-init` docker image (cf the IMAGE_INIT environment variable)

**Review by benoitf — APPROVED** [review]

File: dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/deployment-config.yaml
Comment (davidfestal) [reply dockerfiles/init/modules/openshift/files/scripts/multi-user/keycloak/deployment-config.yaml]: I just made these settings consistent with Keycloak official requirements defined [here](http://www.keycloak.org/docs/latest/server_installation/topics/installation/system-requirements.html).

So I assume we don't need to make this configurable, and can keep it simple.

**Comment by davidfestal** [comment]
ci-build

**Comment by davidfestal** [comment]
@riuvshin @l0rd @skabashnyuk I just submitted a CI build and will wait for the result to consider this as ready for merge on my side. But apart from this, is the PR OK to be merged for you now ?

**Review by l0rd — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Review by riuvshin — APPROVED** [review]

**Comment by skabashnyuk** [comment]
@davidfestal  merge pr when you are ready.

**Pull request merged**
